company.el

```diff
@@ -1218,6 +1223,9 @@ can retrieve meta-data for them."
             (tick (buffer-chars-modified-tick))
             (pt (point))
             (backend company-backend))
+        ;; Indicate that we are currently fetching this result so we do not make
+        ;; another async call again unless we need to.
+        (setq company-candidates-cache (list (cons prefix 'async)))
```

Comment:
No reason to keep the previous entries: if any of them were usable, the async call wouldn't have happened. And filling in cache for prefixes longer than the current is fast enough.